File: daidai-app/app/src/main/java/com/daidai/app/data/root/MagiskHelper.kt
package com.daidai.app.data.root

import java.io.File

/**
 * Magisk目录访问助手
 * Magisk模块通常位于 /data/adb/modules/ 目录下
 */
object MagiskHelper {
    
    // Magisk模块目录
    private const val MAGISK_MODULES_DIR = "/data/adb/modules"
    private const val MAGISK_DATA_DIR = "/data/adb"
    
    // 呆呆面板模块目录
    private const val DAIDAI_MODULE_DIR = "$MAGISK_MODULES_DIR/daidai-panel"
    private const val DAIDAI_DATA_DIR = "/data/local/daidai"
    
    /**
     * 检查是否安装了呆呆面板Magisk模块
     */
    fun isDaidaiModuleInstalled(): Boolean {
        return if (RootChecker.isRooted()) {
            val result = RootChecker.executeAsRoot("test -d $DAIDAI_MODULE_DIR && echo 'exists'")
            result.isSuccess && result.getOrNull()?.contains("exists") == true
        } else {
            false
        }
    }
    
    /**
     * 获取呆呆面板模块信息
     */
    fun getModuleInfo(): Result<MagiskModuleInfo> {
        if (!RootChecker.isRooted()) {
            return Result.failure(Exception("设备未获取root权限"))
        }
        
        return try {
            val moduleProp = RootChecker.readFileAsRoot("$DAIDAI_MODULE_DIR/module.prop")
            if (moduleProp.isFailure) {
                return Result.failure(Exception("无法读取模块信息"))
            }
            
            val props = moduleProp.getOrNull()?.lines()?.associate { line ->
                val parts = line.split("=", limit = 2)
                if (parts.size == 2) parts[0].trim() to parts[1].trim() else "" to ""
            } ?: emptyMap()
            
            Result.success(
                MagiskModuleInfo(
                    id = props["id"] ?: "daidai-panel",
                    name = props["name"] ?: "呆呆面板",
                    version = props["version"] ?: "unknown",
                    versionCode = props["versionCode"]?.toIntOrNull() ?: 0,
                    author = props["author"] ?: "unknown",
                    description = props["description"] ?: ""
                )
            )
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    /**
     * 获取呆呆面板数据目录
     */
    fun getDaidaiDataDir(): String = DAIDAI_DATA_DIR
    
    /**
     * 获取面板配置文件路径
     */
    fun getPanelConfigPath(): String = "$DAIDAI_DATA_DIR/app/Dumb-Panel/config.json"
    
    /**
     * 获取面板数据库路径
     */
    fun getPanelDbPath(): String = "$DAIDAI_DATA_DIR/app/Dumb-Panel/database.db"
    
    /**
     * 获取面板日志目录
     */
    fun getPanelLogsDir(): String = "$DAIDAI_DATA_DIR/app/Dumb-Panel/logs"
    
    /**
     * 获取面板脚本目录
     */
    fun getPanelScriptsDir(): String = "$DAIDAI_DATA_DIR/app/Dumb-Panel/scripts"
    
    /**
     * 读取面板配置
     */
    fun readPanelConfig(): Result<String> {
        if (!RootChecker.isRooted()) {
            return Result.failure(Exception("设备未获取root权限"))
        }
        return RootChecker.readFileAsRoot(getPanelConfigPath())
    }
    
    /**
     * 获取面板端口配置
     */
    fun getPanelPort(): Result<Int> {
        if (!RootChecker.isRooted()) {
            return Result.failure(Exception("设备未获取root权限"))
        }
        
        val portsConf = RootChecker.readFileAsRoot("/data/adb/daidai-panel/ports.conf")
        return portsConf.mapCatching { content ->
            val portLine = content.lines().find { it.startsWith("PANEL_PORT=") }
            portLine?.split("=")?.get(1)?.trim()?.toIntOrNull() ?: 5700
        }
    }
    
    /**
     * 获取所有已安装的Magisk模块
     */
    fun getInstalledModules(): Result<List<MagiskModuleInfo>> {
        if (!RootChecker.isRooted()) {
            return Result.failure(Exception("设备未获取root权限"))
        }
        
        return try {
            val modulesDir = RootChecker.listDirectoryAsRoot(MAGISK_MODULES_DIR)
            if (modulesDir.isFailure) {
                return Result.failure(Exception("无法列出模块目录"))
            }
            
            val modules = modulesDir.getOrNull()
                ?.filter { !it.startsWith(".") && !it.startsWith("total") }
                ?.mapNotNull { line ->
                    val parts = line.split("\\s+".toRegex())
                    if (parts.size >= 9) {
                        val moduleName = parts.last()
                        val moduleProp = RootChecker.readFileAsRoot("$MAGISK_MODULES_DIR/$moduleName/module.prop")
                        if (moduleProp.isSuccess) {
                            val props = moduleProp.getOrNull()?.lines()?.associate { l ->
                                val p = l.split("=", limit = 2)
                                if (p.size == 2) p[0].trim() to p[1].trim() else "" to ""
                            } ?: emptyMap()
                            
                            MagiskModuleInfo(
                                id = props["id"] ?: moduleName,
                                name = props["name"] ?: moduleName,
                                version = props["version"] ?: "unknown",
                                versionCode = props["versionCode"]?.toIntOrNull() ?: 0,
                                author = props["author"] ?: "unknown",
                                description = props["description"] ?: ""
                            )
                        } else null
                    } else null
                } ?: emptyList()
            
            Result.success(modules)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
}

/**
 * Magisk模块信息
 */
data class MagiskModuleInfo(
    val id: String,
    val name: String,
    val version: String,
    val versionCode: Int,
    val author: String,
    val description: String
)

File: lib/services/root/magisk_helper.dart
import 'root_service.dart';

class MagiskModuleInfo {
  final String id;
  final String name;
  final String version;
  final int versionCode;
  final String author;
  final String description;

  MagiskModuleInfo({
    required this.id,
    required this.name,
    required this.version,
    required this.versionCode,
    required this.author,
    required this.description,
  });

  factory MagiskModuleInfo.fromMap(Map<String, dynamic> map) {
    return MagiskModuleInfo(
      id: map['id'] ?? 'unknown',
      name: map['name'] ?? 'unknown',
      version: map['version'] ?? 'unknown',
      versionCode: map['versionCode'] ?? 0,
      author: map['author'] ?? 'unknown',
      description: map['description'] ?? '',
    );
  }
}

class MagiskHelper {
  static const String _magiskModulesDir = '/data/adb/modules';
  static const String _magiskDataDir = '/data/adb';
  static const String _daidaiModuleDir = '$_magiskModulesDir/daidai-panel';
  static const String _daidaiDataDir = '/data/local/daidai';
  
  static Future<bool> isDaidaiModuleInstalled() async {
    if (!await RootService.isRooted()) return false;
    
    try {
      final result = await RootService.executeAsRoot(
        'test -d $_daidaiModuleDir && echo "exists"',
      );
      return result.contains('exists');
    } catch (e) {
      return false;
    }
  }
  
  static Future<MagiskModuleInfo?> getModuleInfo() async {
    if (!await RootService.isRooted()) return null;
    
    try {
      final content = await RootService.readFileAsRoot('$_daidaiModuleDir/module.prop');
      final props = <String, String>{};
      
      for (final line in content.split('\n')) {
        final parts = line.split('=');
        if (parts.length >= 2) {
          props[parts[0].trim()] = parts.sublist(1).join('=').trim();
        }
      }
      
      return MagiskModuleInfo(
        id: props['id'] ?? 'daidai-panel',
        name: props['name'] ?? '呆呆面板',
        version: props['version'] ?? 'unknown',
        versionCode: int.tryParse(props['versionCode'] ?? '0') ?? 0,
        author: props['author'] ?? 'unknown',
        description: props['description'] ?? '',
      );
    } catch (e) {
      return null;
    }
  }
  
  static String getDaidaiDataDir() => _daidaiDataDir;
  
  static String getPanelConfigPath() => '$_daidaiDataDir/app/Dumb-Panel/config.json';
  
  static String getPanelDbPath() => '$_daidaiDataDir/app/Dumb-Panel/database.db';
  
  static String getPanelLogsDir() => '$_daidaiDataDir/app/Dumb-Panel/logs';
  
  static String getPanelScriptsDir() => '$_daidaiDataDir/app/Dumb-Panel/scripts';
  
  static Future<String?> readPanelConfig() async {
    if (!await RootService.isRooted()) return null;
    
    try {
      return await RootService.readFileAsRoot(getPanelConfigPath());
    } catch (e) {
      return null;
    }
  }
  
  static Future<int> getPanelPort() async {
    if (!await RootService.isRooted()) return 5700;
    
    try {
      final content = await RootService.readFileAsRoot('/data/adb/daidai-panel/ports.conf');
      for (final line in content.split('\n')) {
        if (line.startsWith('PANEL_PORT=')) {
          return int.tryParse(line.split('=')[1].trim()) ?? 5700;
        }
      }
      return 5700;
    } catch (e) {
      return 5700;
    }
  }
  
  static Future<List<MagiskModuleInfo>> getInstalledModules() async {
    if (!await RootService.isRooted()) return [];
    
    try {
      final entries = await RootService.listDirectoryAsRoot(_magiskModulesDir);
      final modules = <MagiskModuleInfo>[];
      
      for (final entry in entries) {
        if (entry.startsWith('.') || entry.startsWith('total')) continue;
        
        final parts = entry.split(RegExp(r'\s+'));
        if (parts.length >= 9) {
          final moduleName = parts.last;
          
          try {
            final content = await RootService.readFileAsRoot(
              '$_magiskModulesDir/$moduleName/module.prop',
            );
            final props = <String, String>{};
            
            for (final line in content.split('\n')) {
              final p = line.split('=');
              if (p.length >= 2) {
                props[p[0].trim()] = p.sublist(1).join('=').trim();
              }
            }
            
            modules.add(MagiskModuleInfo(
              id: props['id'] ?? moduleName,
              name: props['name'] ?? moduleName,
              version: props['version'] ?? 'unknown',
              versionCode: int.tryParse(props['versionCode'] ?? '0') ?? 0,
              author: props['author'] ?? 'unknown',
              description: props['description'] ?? '',
            ));
          } catch (e) {
            // Skip modules that can't be read
          }
        }
      }
      
      return modules;
    } catch (e) {
      return [];
    }
  }
  
  static Future<Map<String, dynamic>> getSystemInfoViaRoot() async {
    if (!await RootService.isRooted()) return {};
    
    try {
      final results = await Future.wait([
        RootService.executeAsRoot('cat /proc/meminfo | head -3'),
        RootService.executeAsRoot('cat /proc/cpuinfo | grep "model name" | head -1'),
        RootService.executeAsRoot('df -h / | tail -1'),
        RootService.executeAsRoot('uptime'),
      ]);
      
      return {
        'memory': results[0],
        'cpu': results[1],
        'disk': results[2],
        'uptime': results[3],
        'source': 'root',
      };
    } catch (e) {
      return {};
    }
  }
  
  static Future<String> getPanelLogsViaRoot({int lines = 100}) async {
    if (!await RootService.isRooted()) return '';
    
    try {
      return await RootService.executeAsRoot(
        'tail -n $lines ${getPanelLogsDir()}/*.log 2>/dev/null || echo "No logs found"',
      );
    } catch (e) {
      return '';
    }
  }
}
